src/rgw/rgw_lc.cc

```diff
@@ -590,8 +598,42 @@ static int remove_expired_obj(const DoutPrefixProvider *dpp, lc_op_ctx& oc, bool
   ACLOwner bucket_owner;
   bucket_owner.set_id(bucket_info.owner);
 
-  return obj->delete_object(dpp, &oc.rctx, obj_owner, bucket_owner, meta.mtime, false, 0,
-			    version_id, null_yield);
+  if (oc.cct->_conf->rgw_lc_notify) {
+    notify_res = rgw::notify::reservation_t(
+      store, dpp, &oc.rctx, obj.get(),
+      bucket.get(), lc_id, const_cast<std::string&>(oc.bucket->get_tenant()),
+      lc_req_id, null_yield);
+
+    /* XXX do we agree that lifecycle expiration can be allowed to
+     * fail when notifications are blocked?
+     */
+    ret = rgw::notify::publish_reserve(event_type, *notify_res, nullptr);
+    if (ret < 0) {
+      ldpp_dout(dpp, 1)
+	<< "ERROR: notify reservation failed, deferring delete of object k="
+	<< o.key
+	<< dendl;
+      return ret;
+    }
+  }
+
+  ret = obj->delete_object(dpp, &oc.rctx, obj_owner, bucket_owner, meta.mtime,
```

Comment:
if operation fails, we should not "publish_commit"

Reply:
sorry, should be in the condition